src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,89 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
+        {
+            int length;
+            byte[] jsonBytes;
+            var sharedArrayPool = ArrayPool<byte>.Shared;
+            bool wasRented = false;
+            if (contentLength.HasValue)
+            {
+                length = (int)contentLength;
+                jsonBytes = sharedArrayPool.Rent(length);
+                wasRented = true;
+                await stream.ReadAsync(jsonBytes, 0, length);
+            }
+            else
+            {
+                using var ms = new MemoryStream();
+                await stream.CopyToAsync(ms).ConfigureAwait(false);
+                jsonBytes = ms.ToArray();
+                length = jsonBytes.Length;
+            }
+            try
+            {
+                return Process(jsonBytes, length);
+            }
+            finally
+            {
+                if (wasRented)
+                {
+                    sharedArrayPool.Return(jsonBytes);
+                }
+            }
+        }
+
+        private GraphQLRequestDeserializationResult Process(byte[] jsonBytes, int length)
+        {
+            var jsonData = jsonBytes.AsSpan(0, length);
+
+            JsonTokenType tokenType;
+            try
+            {
+                tokenType = GetTokenType(jsonData);
+            }
+            catch (Exception)
+            {
+                tokenType = JsonTokenType.None;
+            }
+
+            var result = new GraphQLRequestDeserializationResult() { WasSuccessful = true };
+            switch (tokenType)
+            {
+                case JsonTokenType.StartObject:
+                    result.Single = JsonSerializer.Deserialize<GraphQLRequest>(jsonData, _serializerOptions);
+                    break;
+                case JsonTokenType.StartArray:
+                    result.Batch = JsonSerializer.Deserialize<GraphQLRequest[]>(jsonData, _serializerOptions);
+                    break;
+                default:
+                    result.WasSuccessful = false;
+                    break;
+            }
+            return result;
+        }
+
+        private static JsonTokenType GetTokenType(Span<byte> jsonData)
+        {
+            var reader = new Utf8JsonReader(jsonData);
+            var success = reader.Read();
+            return success ? reader.TokenType : JsonTokenType.None;
```

Comment:
I'm not sure if this is the right way to go. It seems silly to create a reader to just Read the first token, but then not re-use that in the caller to actually deserialize the result.

Perhaps @JamesNK could weight in here if he doesn't mind. 

Alternatively, could we just check the raw value of the byte against what a `[` or a `{` is in UTF-8? I'm really not confident in this area. Could it be possible to get a different character encoding at which point the byte value would be different in different scenarios?

Reply:
`Utf8JsonReader` is struct and `reader.Read()` has very low computational cost.

> Alternatively, could we just check the raw value of the byte against what a [ or a { is in UTF-8? 

Of course. If you run the benchmark, then I do not think that you see a significant difference.

Reply:
Sweet, I'm happy if you're happy.